driver-core/src/main/com/mongodb/internal/time/Timeout.java

```diff
@@ -32,110 +42,198 @@
+        });
+        return Collections.min(list, (a, b) -> {
+            TimePoint tpa = (TimePoint) a;
+            TimePoint tpb = (TimePoint) b;
+            return tpa.compareTo(tpb);
+        });
     }
 
     /**
-     * @see TimePoint#isInfinite()
-     */
-    boolean isInfinite();
-
-    /**
-     * @see TimePoint#hasExpired()
-     */
-    boolean hasExpired();
-
-    /**
-     * @see TimePoint#remaining
-     */
-    long remaining(TimeUnit unit);
-
-    /**
-     * A convenience method that handles converting to -1, for integration with older internal APIs.
-     * <p>
-     * Returns an {@linkplain #isInfinite() infinite} timeout if {@code durationNanos} is either negative
-     * or is equal to {@link Long#MAX_VALUE}, otherwise a timeout of {@code durationNanos}.
-     * <p>
-     * Note that the contract of this method is also used in some places to specify the behavior of methods that accept
-     * {@code (long timeout, TimeUnit unit)}, e.g., {@link com.mongodb.internal.connection.ConcurrentPool#get(long, TimeUnit)},
-     * so it cannot be changed without updating those methods.</p>
+     * @return an infinite (non-expiring) timeout
      */
-    default long remainingOrNegativeForInfinite(final TimeUnit unit) {
-        return isInfinite() ? -1 : remaining(unit);
+    static Timeout infinite() {
+        return TimePoint.infinite();
     }
 
     /**
-     * @return an infinite (non-expiring) timeout
+     * @param timeout the timeout
+     * @return the provided timeout, or an infinite timeout if provided null.
      */
-    static Timeout infinite() {
-        return TimePoint.infinite();
+    static Timeout nullAsInfinite(@Nullable final Timeout timeout) {
+        return timeout == null ? infinite() : timeout;
     }
 
     /**
-     * @param duration the positive duration, in the specified time unit.
-     * @param unit the time unit.
+     * @param duration the non-negative duration, in the specified time unit
+     * @param unit the time unit
+     * @param zeroDurationIs what to interpret a 0 duration as (infinite or expired)
      * @return a timeout that expires in the specified duration after now.
      */
-    static Timeout expiresIn(final long duration, final TimeUnit unit) {
+    @NotNull
+    static Timeout expiresIn(final long duration, final TimeUnit unit, final ZeroDurationIs zeroDurationIs) {
         // TODO (CSOT) confirm that all usages in final PR always supply a non-negative duration
         if (duration < 0) {
             throw new AssertionError("Timeouts must not be in the past");
         }
-        return TimePoint.now().timeoutAfterOrInfiniteIfNegative(duration, unit);
+
+        if (zeroDurationIs == ZeroDurationIs.INFINITE) {
+            if (duration == 0) {
+                return Timeout.infinite();
+            }
+            return expiresIn(duration, unit, ZeroDurationIs.EXPIRED);
+        } else {
+            return TimePoint.now().timeoutAfterOrInfiniteIfNegative(duration, unit);
```

Comment:
That is just the way to create a timeout. The extra "OrInfiniteIfNegative" wording is meant to remind of the semantics (which seems more important than possible confusion, as has happened here), rather than suggest that there is an alternate "timeoutAfter" method that should be used. I have added a comment.